File: code-awesome/ruby-on-rails-react/webshop/backend/src/pages/product..rb
# prg3/ruby-on-rails-react/webshop/backend/src/pages/product.rb
# Product Page - Displays a product and links to cart and checkout.

class ProductPage
  def initialize(product_data)
    @product = product_data
    @cart = Cart.new
    @checkout = CheckoutPage.new
  end

  def render
    content = <<~HTML
      <h1>#{@product.name}</h1>
      <img src="#{URLHelper.image_url(@product.image)}" alt="#{URLHelper.image_alt_text(@product.image)}" width="300">
      <p>Price: $#{@product.price}</p>
      <p>Description: #{@product.description}</p>
      <button onclick="location.href='#{URLHelper.cart_url}'" >Add to Cart</button>
      <button onclick="location.href='#{URLHelper.checkout_url}'" >Checkout</button>
    HTML
    return content
  end
end

# Helper Classes for URL Generation and Cart/Checkout Interaction
class URLHelper
  def self.image_url(image_path)
    "#{base_url}/images/#{image_path}"
  end

  def self.image_alt_text(image_path)
    "Image of #{image_path}"
  end

  def self.cart_url
    "/cart"
  end

  def self.checkout_url
    "/checkout"
  end

  def self.base_url
    "/webshop"
  end
end

class Cart
  def initialize
    # Placeholder for cart logic.  Currently, it does nothing.
  end

  def add_to_cart(product_id)
    # In a real implementation, this would update the cart data.
    puts "Adding product #{product_id} to cart."
  end
end

class CheckoutPage
  def initialize
    # Placeholder for checkout logic. Currently, it does nothing.
  end
end

File: code-awesome/django-react/webshop/frontend/src/pages/product..js
import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { useInject } from 'react-use';

const ProductPage = () => {
  const navigate = useNavigate();
  const [product, setProduct] = useState(null);
  const [errorCount, setErrorCount] = useState(0);

  useInject([{
    name: 'productData',
    deps: ['product']
  }])

  useEffect(() => {
    const fetchData = async () => {
      try {
        const data = await productData;
        setProduct(data);
        setErrorCount(0);
      } catch (error) {
        console.error("Error fetching product data:", error);
        setErrorCount(1);
      }
    };

    fetchData();
  }, []);

  const handleClick = () => {
    navigate('/cart');
  };

  const handleClickCheckout = () => {
    navigate('/checkout');
  };

  if (!product) {
    return <div>Loading product...</div>;
  }

  return (
    <div>
      <h2>{product.name}</h2>
      <img src={product.image} alt={product.name} />
      <p>{product.description}</p>
      <p>Price: ${product.price}</p>
      <button onClick={handleClick}>Add to Cart</button>
      <button onClick={handleClickCheckout}>Checkout</button>
    </div>
  );
};

export default ProductPage;